# **콘텐츠 기반 필터링 - part 2**


: 추천의 기준은 모두 콘텐츠에 있다는 점이 핵심

: 그렇기 때문에 영화가 가진 다양한 특징, 정보간의 유사성을 분석하고 이를 기반으로 추천하는 시스템으로 구현된다

ex . 감독, 배우, 내용, 장르, 키워드, 스텝 등등


1. 필요한 library들 불러오기

In [ ]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet


import warnings; warnings.simplefilter('ignore')

2. 사용하려는 데이터 소개, 데이터 전처리

In [ ]:
md = pd. read_csv('/content/drive/MyDrive/movies_metadata.csv')
md

adult  ... vote_count
0      False  ...     5415.0
1      False  ...     2413.0
2      False  ...       92.0
3      False  ...       34.0
4      False  ...      173.0
...      ...  ...        ...
45461  False  ...        1.0
45462  False  ...        3.0
45463  False  ...        6.0
45464  False  ...        0.0
45465  False  ...        0.0

[45466 rows x 24 columns]

MovieLens dataset 중 100,000 ratings, 1,300 tag applications applied to 9000 movies by 700 users

In [ ]:
links_small = pd.read_csv('/content/drive/MyDrive/links_small.csv')
links_small

movieId   imdbId    tmdbId
0           1   114709     862.0
1           2   113497    8844.0
2           3   113228   15602.0
3           4   114885   31357.0
4           5   113041   11862.0
...       ...      ...       ...
9120   162672  3859980  402672.0
9121   163056  4262980  315011.0
9122   163949  2531318  391698.0
9123   164977    27660  137608.0
9124   164979  3447228  410803.0

[9125 rows x 3 columns]

imdld : IMDB(영화 정보 모음 사이트) 의 정보를 포함

tmbld : TMDB(영화 정보 사이트) 의 정보를 포함


In [ ]:
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')

In [ ]:
md = md.drop([19730, 29503, 35587])
#Check EDA Notebook for how and why I got these indices.
md['id'] = md['id'].astype('int')

In [ ]:
# md 데이터와 links_small 데이터를 동일한 'id'값을 이용하여 merge하는 과정
# 이 과정에서 생성된 새로운 데이터 셋을 smd라는 변수에 대입 

smd = md[md['id'].isin(links_small)]
smd.shape

(9099, 24)

총 9099개의 movie에 대한 정보를 담고 있음을 확인할 수 있다.

# Movie Description Based Recommender

1.   movie description을 벡터화 시키기



활용할 데이터 : tagline, overview 열에 있는 데이터

---
아이템 프로필 행렬 : 영화와 영화 정보 간의 관계가 '숫자'형식, 점수 형식으로 나타나 있어야 함

그러나 tagline, overview 열은 단순히 영화의 내용들을 'vector', 'list' 형태로 담고 있기 때문에, 이를 TF-IDF 기법을 통해서 tagline, overview column 에 있는 단어들이 그 영화의 특성을 나타내는데 얼마나 큰 영향을 미치는지를 수치화하여 나타내는 작업이 필요

-> 이 과정을 통해 아이템 프로필 생성 가능



In [ ]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [ ]:
tf = TfidfVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tf.fit_transform(smd['description'])


1) TfidVectorizer 함수  = 단어의 개수 그대로를 카운트 하지 않고, 모든 문서에 공통적으로 들어있는 단어의 경우 문서 구별 
능력이 떨어진다고 보아 가중치를 축소하여 점수를 매기는 함수

tfidf(d,t) = tf(d,t) * idf(t)

tf : 특정한 단어의 빈도수

idf : 특정한 단어가 들어 있는 문서의 수에 반비례하는 수


2) 함수 내의 option

analyzer = 'word' : word 문자열을 기준으로 분석

ngram_range : 단어장 생성에 사용할 토큰의 크기 결정

min_df, max_df : 문서에서 토큰이 나타난 횟수를 기준으로 단어장 구성하도록 하는 옵션
, 토큰의 빈도가 min과 max 사이에 포함되어 있어야만 단어장에 추가

stop_words = 'english'

: 단어장을 생성할 때 무시할 수 있는 단어

: a, the와 같은 불용어 제거하는 과정


In [ ]:
tfidf_matrix.shape

(9099, 268124)

**2. Cosine Similarity를 활용하여 두 영화간의 유사성을 계산하기**

cosine(x, y) =  (x.y⊺) / (|| x ||. || y ||)

: TF-IDF Vectorizer 과 dot product를 이용하면 코사인 유사도 점수가 제공됨


In [ ]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [ ]:
cosine_sim[0]

array([1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
       0.        ])

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [ ]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

In [ ]:
get_recommendations('The Dark Knight').head(10)

7931                      The Dark Knight Rises
132                              Batman Forever
1113                             Batman Returns
8227    Batman: The Dark Knight Returns, Part 2
7565                 Batman: Under the Red Hood
524                                      Batman
7901                           Batman: Year One
2579               Batman: Mask of the Phantasm
2696                                        JFK
8165    Batman: The Dark Knight Returns, Part 1
Name: title, dtype: object

# Metadata Based Recommender

In [ ]:
credits = pd.read_csv('/content/drive/MyDrive/credits.csv')
keywords =pd.read_csv('/content/drive/MyDrive/keywords.csv')

In [ ]:
# casting 정보, crew 정보
credits


cast  ...      id
0      [{'cast_id': 14, 'character': 'Woody (voice)',...  ...     862
1      [{'cast_id': 1, 'character': 'Alan Parrish', '...  ...    8844
2      [{'cast_id': 2, 'character': 'Max Goldman', 'c...  ...   15602
3      [{'cast_id': 1, 'character': "Savannah 'Vannah...  ...   31357
4      [{'cast_id': 1, 'character': 'George Banks', '...  ...   11862
...                                                  ...  ...     ...
45471  [{'cast_id': 0, 'character': '', 'credit_id': ...  ...  439050
45472  [{'cast_id': 1002, 'character': 'Sister Angela...  ...  111109
45473  [{'cast_id': 6, 'character': 'Emily Shaw', 'cr...  ...   67758
45474  [{'cast_id': 2, 'character': '', 'credit_id': ...  ...  227506
45475                                                 []  ...  461257

[45476 rows x 3 columns]

In [ ]:
#keyword 정보
keywords


id                                           keywords
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...
...       ...                                                ...
46414  439050             [{'id': 10703, 'name': 'tragic love'}]
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...
46416   67758                                                 []
46417  227506                                                 []
46418  461257                                                 []

[46419 rows x 2 columns]

In [ ]:
# int 형태로 변환
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')

In [ ]:
md.shape

(45463, 24)

In [ ]:
#id 값을 기준으로 credits data 와 keywords data를 md 에 merge
md = md.merge(credits, on='id')
md = md.merge(keywords, on='id')

In [ ]:
# md와 line_small data에 동시에 존재하는 movie id만 smd에 넣어준다.
smd = md[md['id'].isin(links_small)]
smd.shape

(9219, 27)

In [ ]:
# cast 데이터에서 ''문자열 안에 있는 document만을 포함시킨다.
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x)) # 배우의 수
smd['crew_size'] = smd['crew'].apply(lambda x: len(x)) # 스텝의 수

In [ ]:
# 스텝중에서도 감독만 추출하는 사용자 정의 함수 생성
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [ ]:
#smd 에 감독 데이터 추가
smd['director'] = smd['crew'].apply(get_director)

In [ ]:
#smd에 배우 이름 추가
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
# 그 중에서도 가장 영향력 있는 세명의 배우만 추가
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)

In [ ]:
# smd에 키워드 데이터 추가
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])


목표 : 장르, 감독, 주인공, 키워드로 이루어진 모든 영화에 대한 meta data dump를 만들기

그 후 countvectorizer을 이용하여 다른 영화들과 얼마나 비슷한지 비교하고 점수를 기준으로 정렬하기

(1) 감독, 스텝 데이터 전처리



In [ ]:
# 모든 데이터 소문자화, 띄어쓰기 없애기
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [ ]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

(2) keyword 데이터 전처리

In [ ]:
# keywords라는 인덱스를 기준으로 각 영화마다의 장르를 동일한 인덱스로 표현
# keyword가 유사하다는 것은 곧 영화의 장르가 비슷하다는 것을 의미하기 때문
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [ ]:
s = s.value_counts()
s[:5]

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: keyword, dtype: int64

In [ ]:
# 개수가 하나뿐인 키워드는 영향력티이 적으므로 없애준다
s = s[s > 1]

In [ ]:
# SnowballStemmer : 복수형, dogs와 같은 단어를 dog로, 즉 원형으로 바꾸어줌
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [ ]:
# 키워드를 골라내서 index에 넣어주는 사용자 정의 함수 생성
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words


In [ ]:
#위에서 생성한 함수를 이용해서 keywords에 있는 데이터 중 키워드를 골라내기
smd['keywords'] = smd['keywords'].apply(filter_keywords)
# keyword의 단어들을 원형으로 바꿔주고, 띄어쓰기를 없앤다
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])


In [ ]:
smd

adult  ...                                           director
0      False  ...         [johnlasseter, johnlasseter, johnlasseter]
1      False  ...            [joejohnston, joejohnston, joejohnston]
2      False  ...         [howarddeutch, howarddeutch, howarddeutch]
3      False  ...   [forestwhitaker, forestwhitaker, forestwhitaker]
4      False  ...         [charlesshyer, charlesshyer, charlesshyer]
...      ...  ...                                                ...
40952  False  ...      [greggchampion, greggchampion, greggchampion]
41172  False  ...  [tinusureshdesai, tinusureshdesai, tinusureshd...
41225  False  ...  [ashutoshgowariker, ashutoshgowariker, ashutos...
41391  False  ...            [hideakianno, hideakianno, hideakianno]
41669  False  ...                  [ronhoward, ronhoward, ronhoward]

[9219 rows x 30 columns]

In [ ]:
#soup 이라는 변수 만들기 : 앞서 생성한 keyword, cast, director에 대한 정보를 다 담고 있는 변수
smd['soup'] = smd['keywords'] + smd['cast'] +smd['director']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))



In [ ]:
# 단어별로 벡터화 시켜주기, 앞선 내용과 유사,
# CountVectorizer함수를 이용해서 단어를 벡터화
# (1) 문서를 토큰 리스트로 변환 (2) 각 문서에서 토큰의 출현 빈도를 센다
# (3) 각 문서를 BOW 인코딩 벡터로 변환한다.
count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [ ]:
# 코사인 유사도 점수 구하기(평가척도)
cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [ ]:
# 영화 인덱스 변수 만들기
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [ ]:
# 앞서 활용했던 사용자 정의 함수
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

In [ ]:
get_recommendations('The Dark Knight').head(10)

8031         The Dark Knight Rises
6218                 Batman Begins
6623                  The Prestige
7648                     Inception
2085                     Following
4145                      Insomnia
3381                       Memento
8613                  Interstellar
7659    Batman: Under the Red Hood
1134                Batman Returns
Name: title, dtype: object

감독, 주연 배우, 키워드 등의 영화 콘텐츠를 기준으로 추가하여 유사한 영화를 추천해주는 improved recommendation system 을 생성하게 됨.